### PR TITLE
Split Local and owned Avatar Entity scripts into their own ScriptEngine

### PR DESCRIPTION
**Pull request by HifiExperiments**
https://github.com/kasenvr/project-athena/issues/655

Test script: https://cdn.vircadia.com/test/scripting/still_running.js

### CODE REVIEW

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR659/Vircadia-Alpha-PR659-8804e9c_2a1857e9.exe

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR659/Vircadia-Alpha-PR659-8804e9c_2a1857e9.exe

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR659/Vircadia-Alpha-PR659-b83de83_2e4a6ee1.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR659/Vircadia-Alpha-PR659-b83de83.tar_2e4a6ee1.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR659/Vircadia-Alpha-PR659-b83de83.tar_2e4a6ee1.Z

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR659/Vircadia-Alpha-PR659-b83de83_2e4a6ee1.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR659/Vircadia-Alpha-PR659-b83de83_2e4a6ee1.exe

**Comment by daleglass** [comment]
It would be nice to have a description of what this does and why

**Comment by HifiExperiments** [comment]
@daleglass Kalila has more details about this + a test case for it, or I can provide more details over PM

but basically, before, all entity scripts ran in a single script engine.  this splits them so that local + owned avatar entities, which persist across domains, are in one script engine, while domain + non-owned avatar entities are in another.  this lets us do some additional needed cleanup on domain switches

File: libraries/entities-renderer/src/EntityTreeRenderer.cpp [hunk 1]
```diff
@@ -158,79 +158,86 @@ render::ItemID EntityTreeRenderer::renderableIdForEntityId(const EntityItemID& i
...
+        foreach (const EntityItemID& entityID, entitiesWithEntityScripts) {
             EntityItemPointer entityItem = getTree()->findEntityByEntityItemID(entityID);
-
             if (entityItem && !entityItem->getScript().isEmpty()) {
                 if (!(entityItem->isLocalEntity() || entityItem->isMyAvatarEntity())) {
-                    if (_currentEntitiesInside.contains(entityID)) {
-                        _entitiesScriptEngine->callEntityScriptMethod(entityID, "leaveEntity");
-                    }
-                    _entitiesScriptEngine->unloadEntityScript(entityID, true);
+                    _nonPersistentEntitiesScriptEngine->unloadEntityScript(entityID, true);
```
Comment (ctrlaltdavid) [inline libraries/entities-renderer/src/EntityTreeRenderer.cpp:240]: Is it valid to remove the `_currentEntitiesInside` test here? Is `_currententitiesIside` still used OK elsewhere?

**Review by ctrlaltdavid — CHANGES_REQUESTED** [review]

File: libraries/entities-renderer/src/EntityTreeRenderer.cpp (hunk 1 above)
Comment (HifiExperiments) [reply libraries/entities-renderer/src/EntityTreeRenderer.cpp:240]: leaveDomainAndNonOwnedEntities already calls leaveEntity on the things we are inside so we shouldn’t need to do it here

**Review by ctrlaltdavid — APPROVED** [review]

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR659/Vircadia-Alpha-PR659-b6edc70.tar_2e4a6ee1.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR659/Vircadia-Alpha-PR659-b6edc70_2e4a6ee1.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR659/Vircadia-Alpha-PR659-b6edc70.tar_2e4a6ee1.Z

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR659/Vircadia-Alpha-PR659-b6edc70_2e4a6ee1.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR659/Vircadia-Alpha-PR659-b6edc70_2e4a6ee1.exe

**Comment by kasenvr** [comment]
I will create a test script for this.

**Comment by kasenvr** [comment]
Brought up by Vegaslon: Will this work with https://apidocs.vircadia.dev/Entities.html#.callEntityMethod still?

**Comment by kasenvr** [comment]
Test script: https://cdn.vircadia.com/test/scripting/still_running.js

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR659/Vircadia-Alpha-PR659-c640974_2e4a6ee1.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR659/Vircadia-Alpha-PR659-c640974_2e4a6ee1.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR659/Vircadia-Alpha-PR659-c640974.tar_2e4a6ee1.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR659/Vircadia-Alpha-PR659-c640974.tar_2e4a6ee1.gz

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR659/Vircadia-Alpha-PR659-c640974_2e4a6ee1.exe

**Review by kasenvr — APPROVED** [review]
Worked fine for me under extended use, my test script also worked. :)

**Comment by kasenvr** [comment]
Tested again for myself with Silverfish and all of a sudden... now she's not working, not sure what I'm doing differently this time to test.

Tested in this way: 

1. Create a cube 
2. Attach this script to that cube and confirm some log spam for messages...
3. Delete the cube.
4. Confirm log spam still happens while on the domain.
5. Navigate to a new domain and confirm you no longer can get log spam (a good way to trigger log spam is to open the Create app)

What happens for Silverfish and I when testing most recently is that this script is still running even though we navigated to a new domain.

**Comment by HifiExperiments** [comment]
@ctrlaltdavid @kasenvr I made a few more changes and now hopefully we should be good.  I think this might also fix some of the ScriptEngine crashes that happen when you reload content now

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR659/Vircadia-Alpha-PR659-b232156.tar_3b2c219e.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR659/Vircadia-Alpha-PR659-b232156_3b2c219e.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR659/Vircadia-Alpha-PR659-b232156.tar_3b2c219e.gz

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR659/Vircadia-Alpha-PR659-b232156_3b2c219e.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR659/Vircadia-Alpha-PR659-b232156_3b2c219e.exe

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR659/Vircadia-Alpha-PR659-460bccf_72d43947.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR659/Vircadia-Alpha-PR659-460bccf.tar_72d43947.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR659/Vircadia-Alpha-PR659-460bccf.tar_72d43947.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR659/Vircadia-Alpha-PR659-460bccf_72d43947.sh

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR659/Vircadia-Alpha-PR659-460bccf_72d43947.exe

**Comment by kasenvr** [comment]
The bug appears to be gone! I'll use it like normal for a bit longer to ensure all is how it should be. :)

**Comment by kasenvr** [comment]
OK after using it a bit more extensively... something seems to be wrong with loading domains. A domain will load a little then stop entirely, sometimes showing just the void, and refuse to load.

You can solve this by reloading content in which the domain will actually load completely almost immediately for a few seconds then the content reload will trigger, thereby wiping it all out and then loading it correctly again. Not sure why this is happening. Aitolda and I used Alezia's build PR858 to try and figure out where this might be dying.... this issue was not present on hers (build 17 hours ago from the time of writing this comment.)

**Comment by kasenvr** [comment]
Update: running this PR build again, it appears to be working fine... according to Aitolda it happens to him often/not often at different times on 2020.3.1 as well, so mayhaps that was just a fluke that I was hitting heavily for a bit there.

We may want to find out where that bug is coming from, but I don't believe it is related to this PR at the moment though..

**Comment by HifiExperiments** [comment]
@kasenvr I’ve also been experiencing that issue recently, not just on this PR.  Logs show some issues with “silent domain check ins”.  I wonder if it’s related to the new metaverse?

**Review by ctrlaltdavid — APPROVED** [review]

**Review by kasenvr — APPROVED** [review]
Worked fine for me.

**Pull request merged**